### ISSUE TITLE
Experiment Module: link plot attributes to column and level names in CSV

### ISSUE THREAD

**Issue by aislinnpearson** [Rothamsted-Ecoinformatics/farm_rothamsted#275]
Having had some practice now with the experiment module, I have come to the conclusion that it will always be easiest to upload plot attributes - and after speaking to the grassroots team, I think they have found the same.  

For this reason I would suggest that we revert back to the previous way of doing things where you uploaded a GeoJSON and a separate CSV (sorry @Paul!), where the CSV contains the human readable column levels, and these are converted into data attributes on upload.

The main reason I am suggesting this is because I could then see a way in which we could enable editing for plots: you could upload a csv for the plots you want to edit, on upload be asked to add one more flags and a revision log message, and then the values for these plots are updated, incorporating flags and recorded by a change log assigned both against the experiment area and the individual plots.

In addition, I can see that we could use a similar mechanism to upload files to multi-year experiments (although that would have to be a separate process, for which I will make a separate issue)

I could see a scenario where you want to make more than one change, and we'd have to discuss how you could do that. But this would be a first version for implementing an edit function for the plots, as per #195. I think we would need to discuss a different option for editing columns and column levels.

**Comment by aislinnpearson** [comment]
See this comment here for the final agreed mechanism for editing columns, column levels and plot attributes is here: https://github.com/Rothamsted-Ecoinformatics/farm_rothamsted/issues/276#issuecomment-1359895471

**Comment by aislinnpearson** [comment]
This issue was superseded by #358

**Comment by paul121** [comment]
We also need to change this to link plot attribute column levels by name, not ID.

**Comment by aislinnpearson** [comment]
Hi @paul121 

Here are the example files. I know they uploaded when I converted them to the format used in the current live version, so hopefully it will be okay. In case it helps, this is the live experiment here which uses the existing format: https://rothamstedfarm.farmos.net/plan/34

As mentioned in this [comment here](https://github.com/Rothamsted-Ecoinformatics/farm_rothamsted/issues/358#issuecomment-1450412860) (see the first bullet point under 'plots page') I have included both the plot_number _and_ the plot_id in the GeoJSON. Let me know if you think that works as a concept (ironically the two attributes which relate exclusively to the plot asset, as mentioned in the call today). 

Thanks so much for your time

~~[2023-03_2-10-x_example-upload-files.zip](https://github.com/Rothamsted-Ecoinformatics/farm_rothamsted/files/10912606/2023-03_2-10-x_example-upload-files.zip)~~ Updated below

**Comment by paul121** [comment]
@aislinnpearson the plot attribute levels are consistent and reference the column levels by name. However, the column names in the header of this plot attributes file are inconsistent. Some reference the column name, column abbreviation, or the column ontology name. It would be best if we settle on a single thing to reference. For now I will go with column name but let me know if you would like to use something else.

> I have included both the plot_number and the plot_id in the GeoJSON

I am only using the plot_number to cross reference, but feel free to include the plot_id if you like :+1:

**Comment by paul121** [comment]
Oh also, I just wanted to check on your usage of `nil`...
- Right now I see `nil` is provided as a column level, so it seems it is important that this is explicitly included.
- I also see that `None` is used as a column level for the "Nutrition treatment".
- Finally, we previously decided on a rule that a column level value of `na` would simply not be added to a plot. This is distinct form `nil` or `None` because it is not a column level, and will not appear in the Filter options.

For consistency I wonder if you'd prefer to use `None` instead of `nil`, it seems this may appear nicer in the UI. But it is just another column level so of course you can call it whatever you want!

![Screenshot from 2023-03-07 14-31-03](https://user-images.githubusercontent.com/3116995/223569708-1e8295ec-1493-4d0a-b47d-9994551e8589.png)

**Comment by aislinnpearson** [comment]
Hi @paul121,

> the plot attribute levels are consistent and reference the column levels by name. However, the column names in the header of this plot attributes file are inconsistent. Some reference the column name, column abbreviation, or the column ontology name. It would be best if we settle on a single thing to reference. For now I will go with column name but let me know if you would like to use something else.

Sorry! That was me doing this very quickly yesterday evening. I have double checked them now and I hope they are all correct. I have also used column name - I think that is ideal. Updated files below.

[2023-03_2-10-x_example-upload-files.zip](https://github.com/Rothamsted-Ecoinformatics/farm_rothamsted/files/10922124/2023-03_2-10-x_example-upload-files.zip)

> I am only using the plot_number to cross reference, but feel free to include the plot_id if you like 

Can we make this required and a validation step? The reason I ask is that I can see human error creeping in where the plot_number : plot_id combination in the excel file doesn't match the plot_number in the CSV (it's surprising how easy this is to do, and the error is usually only obvious when you visualise the plots using the GIS files). The scientists will also be working with the numbering system they know (plot_id), and it's pretty crucial the two always match up... 

> Oh also, I just wanted to check on your usage of nil...

So this is actually an interesting use case which demonstrates how useful the NA rule is... Park Grass (one of our classical experiments, similar to Broadbalk) where 'None' means 'no nutrients at all' (and so is a specific treatment, similar to a control if this was a replicated trial) while 'nil' has historically always been used demonstrate that, for a specific nutrient within a treatment, a zero quantity is applied. So 'nil' is equivalent to zero for a particular nutrient, while 'None' is no nutrients at all. 'na' is when a treatment (or column) is not applicable to all plots - for example, guard plots which don't receive treatments. Hoep that helps... But good question. I should clarify that in the documentation.

**Comment by paul121** [comment]
> So 'nil' is equivalent to zero for a particular nutrient, while 'None' is no nutrients at all. 'na' is when a treatment (or column) is not applicable to all plots - for example, guard plots which don't receive treatments. 

Makes sense!

**Comment by paul121** [comment]
> Can we make this required and a validation step? The reason I ask is that I can see human error creeping in where the plot_number : plot_id combination in the excel file doesn't match the plot_number in the CSV

Sure thing. I will work on that

**Comment by paul121** [comment]
> Can we make this required and a validation step? The reason I ask is that I can see human error creeping in where the plot_number : plot_id combination in the excel file doesn't match the plot_number in the CSV

Hey @aislinnpearson, sorry I missed that we actually want to do this validation in the GeoJSON upload. That is actually quite a bit more complicated because we have separated the GeoJSON from the other plot attribute upload. In order to perform this verification I would need to query all existing plots...and if there are thousands of plots this gets really complicated. It's also possible for someone to upload the GeoJSON before uploading plot attributes, before the plots have a `plot_id`.

For sake of time I think we should wait on this. Is that OK? I do have validation in place to ensure that 1) all the plot numbers are included and 2) the plot numbers match up with the existing plot numbers.

**Comment by aislinnpearson** [comment]
Hey @paul121,

> That is actually quite a bit more complicated because we have separated the GeoJSON from the other plot attribute upload. In order to perform this verification I would need to query all existing plots...

Oh of course. That's so obvious! Sorry I didn't think about it. 

I've created an issue so we don't lose track of this: 
- #376
